source/fundamentals/crud/write-operations/upsert.txt

```diff
@@ -4,8 +4,200 @@ Insert or Update in a Single Operation
+
+- the number of documents that matched the query filter
+- the number of documents that got modified 
+- the id of the document that got upserted 
+
+An upsert is an update to a document or insert of a document.  An update
+occurs when there is at least one match to your query filter.  An insert
+occurs when there is no match to your query filter.
+
+To set an upsert:
+~~~~~~~~~~~~~~~~~
+
+Use :java-docs:`UpdateOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOptions.html>`
+as a second parameter to the ``updateOne()`` and ``updateMany()``
+methods.
+
+Use :java-docs:`ReplaceOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOptions.html>`
+as a second parameter to the ``replaceOne()`` method.
+
+The following examples are about a paint store that sells eight different
+colors of paint. The store had their annual online sale resulting in the
+following documents in their ``paint_inventory`` collection: 
+
+.. code-block:: json
+
+    { "_id": 1, "color": "red", "qty": 5 }
+    { "_id": 2, "color": "purple", "qty": 8 }
+    { "_id": 3, "color": "blue", "qty": 0 }
+    { "_id": 4, "color": "white", "qty": 0 }
+    { "_id": 5, "color": "yellow", "qty": 6 }
+    { "_id": 6, "color": "pink", "qty": 0 }
+    { "_id": 7, "color": "green", "qty": 0 }
+    { "_id": 8, "color": "black", "qty": 8 }
+
+
+``updateOne()``
+---------------
+
+The updateOne() method updates the specified fields in a single document
+that matches the query filter. If the query filter matches more than one
+document, the method will update the first occurrence of a match in the
+collection. 
+
+Example
+~~~~~~~
+
+The store is restocking their items.  The first shipment that arrived
+are ten cans of orange paint and we need to update this in the collection.
+
+To update orange, query the ``paint_inventory`` collection where the
+``color`` is ``"orange"``, set the update for ``qty`` to ``10``, set
+``UpdateOptions.upsert()`` to ``true`` and pass the query to the
+``updateOne()`` method: 
+
+.. literalinclude:: /includes/fundamentals/code-snippets/InsertUpdate.java
+   :language: java
+   :dedent:
+   :start-after: begin updateOneExample
+   :end-before: end updateOneExample
+
+The method returns:
+
+.. code-block:: json   
+
+    AcknowledgedUpdateResult{ matchedCount=0, modifiedCount=0, upsertedId=BsonObjectId{ value=606782721601f9443bc1c7a7 }} 
+
+The following shows the documents in the ``paint_inventory`` collection:
+
+.. code-block:: json   
+
+    { "_id": 1, "color": "red", "qty": 5 }
+    { "_id": 2, "color": "purple", "qty": 8 }
+    { "_id": 3, "color": "blue", "qty": 0 }
+    { "_id": 4, "color": "white", "qty": 0 }
+    { "_id": 5, "color": "yellow", "qty": 6 }
+    { "_id": 6, "color": "pink", "qty": 0 }
+    { "_id": 7, "color": "green", "qty": 0 }
+    { "_id": 8, "color": "black", "qty": 8 }
+    { "_id": { "$oid": "606782721601f9443bc1c7a7" }, "color": "orange", "qty": 10 }
```

Comment:
issue: use consistent _id values (oid). It is a serious anti-pattern to mix their types.